Add a basic Sankey Diagram to visualize dependencies

MBuchalik opened the pull request:
This PR adds a Sankey Diagram visualization of our dependencies.

For this, we first take our Service Docs and build something called "Intermediate Graph". This is a data structure that is based on these Service Docs, but with some advantages: All Services, APIs, and Events, are modelled as graph nodes. A Service Node holds references to all API and Event Nodes that are produced/consumed by this service. An API/Event Node holds references to all Services producing/consuming this API/Event. Thus, we are able to easily traverse the graph simply by following references. Also, a "Pivot Service" needs to be provided when generating the Intermediate Graph. This Pivot Service is usually the service the user has currently selected in the UI. In the Intermediate Graph, for every node we compute the distance from this Pivot Service. This makes it super easy to do things like "highlight all Events that are directly connected to the Pivot Service", or "higlight all services that are only two 'hops' away from the Pivot Service".

After the Intermediate Graph has been created, we create the actual Sankey nodes (based on this Intermediate Graph). This is now a pretty easy job, since we have a lot of information about the relationships between the individual Services, APIs, and Events. APIs/Events that are directly connected to the Pivot Service are pink/cyan, while other APIs/Events are grey. Services that are two "hops" away from the Pivot Service are colored with a random but stable color. Other services are grey. The same goes for links: A link between two nodes is grey if one of the nodes is grey. This shows the strengths of the Intermediate Graph: Choosing a color was super easy, because I just had to look up how many "hops" a particular node is away from the Pivot Service.

Technically, we could have done this in one step. However, I believe that the Intermediate Graph makes it easier to add additional visualizations in the future.

There is one probably controversial point. As you can see in the second screenshot, some service nodes are colored, even though they only have grey links. This is actually intended: If a service is 2 hops away from the Pivot Service, it gets a color. Let's say a service is only connected to the Pivot Service using Events. But in the left side bar, the user has decided only to include APIs in the visualization. Then, the node is still colored. I have decided to go for this approach because it makes the implementation much easier. Also, I think this might provide some insights that would otherwise be lost. There is another case where this might happen: If a node appears on the left and on the right side, then the node always gets the same color. So it might happen that the node on the left side has only grey outgoing edges, but the node on the right side has some colored ingoing edges.

# Screenshots

![grafik](https://user-images.githubusercontent.com/8061217/197489427-690317ae-0d6b-4502-b1ae-9442a613b569.png)
![grafik](https://user-images.githubusercontent.com/8061217/197490368-1a142fa1-9dee-4e7e-a552-e13a23cc7fe5.png)

Here, I am hovering "TransformationService":
![grafik](https://user-images.githubusercontent.com/8061217/197491896-ee952ad3-7aff-4312-98e8-b55bc13e8811.png)

Here, I am hovering an edge:
![grafik](https://user-images.githubusercontent.com/8061217/197491964-1c587962-cc6f-426e-923c-c1b04ccbca99.png)


This happens if no data is available for visualization:
![grafik](https://user-images.githubusercontent.com/8061217/197492087-33783ae5-7ad1-43c0-b6a7-cf3fd9d33143.png)


![grafik](https://user-images.githubusercontent.com/8061217/197492137-f9c11c12-9ac7-4d06-a4f5-00d5f95fb096.png)
![grafik](https://user-images.githubusercontent.com/8061217/197492267-d8f4781d-bbbe-47fe-9be2-a125b5dfa919.png)
![grafik](https://user-images.githubusercontent.com/8061217/197492307-3bd23bc3-cd70-4371-85c6-d4b1840b69a0.png)

This might also happen: The WebClient does not consume any Events:

![grafik](https://user-images.githubusercontent.com/8061217/197492425-16ac92ec-f898-482f-8384-ccb5afffec6e.png)

georg-schwarz commented on frontend/src/components/main-page/groups-tree-page/service-details/dependencies/visualization/visualization-modal.tsx:
```diff
@@ -0,0 +1,228 @@
+
+      <Box sx={{ display: 'flex', alignItems: 'center', gap: 2 }}>
+        <span>{props.link.source.label}</span>
+        <Icons.ArrowForward />
```
Can we have different tooltips depending on if event or API?
I could imagine that showing the following would be valuable to users:
* "WebClient consumes API /notifications/configs"
* "NotificationService provides API /notifications/configs"
* "LoadService produces event load.execution.success"
* "NotificationService consumes event load.execution.success"

If we decide to implement something like this it should be a separate issue.

MBuchalik replied:
Yes, that should be possible 👍

georg-schwarz commented:
Really good job!

I made two minor changes:
* Add to the middle node if it is an event or API
* Reversed the presentation of event flows. Now event publishers are left, consumers are right leading to a reading flow from left to right. Let me know if you think that is not intuitive!

![image](https://user-images.githubusercontent.com/28054628/197718055-fee13e96-a62a-4e69-a9e9-0815f80bbf45.png)

@MBuchalik please have a look if I did the changes in the right places.

MBuchalik commented:
@georg-schwarz I have added some minor code style improvements. Also, I have changed the comments to not talk about "left" and "right", because the Intermediate Graph could/will in the future be used for other visualizations (like Networks) where we have no left and right, or where left and right have different semantics. (Also, the two comments now include an example, because I personally needed this to understand the difference between APIs and Events 😄)

From my point of view, this can now be merged 👍

georg-schwarz commented on frontend/src/components/main-page/groups-tree-page/service-details/dependencies/visualization/sankey-utils.ts:
```diff
@@ -0,0 +1,284 @@
+    customData: {
+      label: nodeName,
+      color: getColorForNode(correspondingIntermediateNode),
+      position: nodePosition,
```
Yeah, let's do a separate issue. It is not critical, so we also might just close it if it is not possible - but then we have it documented ;)
